bluemira/equilibria/fem_fixed_boundary/fem_magnetostatic_2D.py

```diff
@@ -0,0 +1,180 @@
+# License along with bluemira; if not, see <https://www.gnu.org/licenses/>.
+
+"""
+Bluemira module for the solution of a 2D magnetostatic problem with cylindrical symmetry
+and toroidal current source using fenics FEM solver
+"""
+
+import dolfin
+
+
+class FemMagnetostatic2d:
+    """
+    A 2D magnetostic solver. The solver is thought as support for the fem fixed
+    boundary module and it is limited to axisymmetric magnetostatic problem
+    with toroidal current sources. The Maxwell equations, as function of the poloidal
+    magnetic flux (:math:`\Psi`), are then reduced to the form ([Zohm]_, page 25):
+
+    .. math::
+        r^2 {\\nabla}{\cdot}{\\left(}{\\frac{{\\nabla}{\psi}}{r^2}}{\\right)} = 2
+        \pi r \mu_0 J_{\Phi}
+
+    whose weak formulation is defined as ([Villone]_):
+
+    .. math::
+        \\int_{D_p} {\\frac{1}{r}}{\\nabla}{\Psi}{\cdot}{\\nabla} v \,dr\,dz = 2
+        \pi r \mu_0 \\int_{D_p} J_{\Phi} v \,dr\,dz
+
+    where :math:`v` is the basis element function of the defined functional subspace
+    :math:`V`.
+
+    .. [Zohm] H. Zohm, Magnetohydrodynamic Stability of Tokamaks, Wiley-VCH, Germany,
+       2015
+    .. [Villone] VILLONE, F. et al. Plasma Phys. Control. Fusion 55 (2013) 095008,
+       https://doi.org/10.1088/0741-3335/55/9/095008
+
+    Parameters
+    ----------
+    mesh : dolfin.mesh or string
+           the filename of the xml file with the mesh definition
+           or a dolfin mesh
+    boundaries : dolfin.MeshFunction or string
+                 the filename of the xml file with the boundaries definition
+                 or a MeshFunction that defines the boundaries
+    p : int
+        the order of the approximating polynomial basis functions
+    """  # noqa (W505)
+
+    def __init__(self, mesh, boundaries=None, p=3):
+        # ======================================================================
+        # define the geometry
+        if isinstance(
+            mesh, str
+        ):  # check wether mesh is a filename or a mesh, then load it or use it
+            self.mesh = dolfin.Mesh(mesh)  # define the mesh
+        else:
+            self.mesh = mesh  # use the mesh
+
+        # ======================================================================
+        # define boundaries
+        if boundaries is None:  # Dirichlet B.C. are defined
+            self.boundaries = dolfin.MeshFunction(
+                "size_t", mesh, mesh.topology().dim() - 1
+            )  # initialize the MeshFunction
+        elif isinstance(boundaries, str):
+            # check wether boundaries is a filename or a MeshFunction,
+            # then load it or use it
+            self.boundaries = dolfin.MeshFunction(
+                "size_t", self.mesh, boundaries
+            )  # define the boundaries
+        else:
+            self.boundaries = boundaries
+
+        # ======================================================================
+        # define the function space and bilinear forms
```

Comment:
Update comment to include explanation for the use of "CG" (continuous galerkin). It's easy to overlook but important

Reply:
comment updated